File: ios/ExperimentReactNativeClient.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(ExperimentReactNativeClient, NSObject)

RCT_EXTERN_METHOD(initialize:(NSString *)apiKey config:(NSDictionary *)config resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(initializeWithAmplitudeAnalytics:(NSString *)apiKey config:(NSDictionary *)config resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(fetch:(NSDictionary *)user resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setUser:(NSDictionary *)user resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(variant:(NSString *)key resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(variantWithFallback:(NSString *)key fallback:(NSDictionary *)fallback resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(all: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setAmplitudeUserProvider:(NSString *)amplitudeInstanceName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/ExperimentReactNativeClient.swift
import Foundation

#if RCT_NEW_ARCH_ENABLED
import RNExperimentReactNativeClientSpec
#endif

@objc(ExperimentReactNativeClient)
class ExperimentReactNativeClient: NSObject {

    private let appleContextProvider = AppleContextProvider()

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    @objc
    func getApplicationContext(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let applicationContext = [
            "version": appleContextProvider.version,
            "platform": appleContextProvider.platform,
            "language": appleContextProvider.language,
            "os": appleContextProvider.os,
            "device_manufacturer": appleContextProvider.deviceManufacturer,
            "device_model": appleContextProvider.deviceModel,
        ]
        resolve(applicationContext)
    }
}

#if RCT_NEW_ARCH_ENABLED
extension ExperimentReactNativeClient: NativeExperimentReactNativeClientSpec {}
#endif
